File: src/Forge.Host/Bridge/RenderPayloadCodec.cs
namespace Forge.Host.Bridge;

public static partial class BridgePayloadCodec
{
    public static byte[] EncodeUyaRenderPackageRequest(UyaRenderPackageRequestPayload value)
    {
        var writer = new PayloadWriter();
        writer.WriteString(value.SourceIsoPath);
        writer.WriteString(value.CacheRootPath);
        writer.WriteString(value.Fingerprint);
        writer.WriteUInt32(value.Level);
        writer.WriteString(value.ProjectPath);
        writer.WriteString(value.CatalogRootPath);
        return writer.ToArray();
    }

    public static UyaRenderPackageRequestPayload DecodeUyaRenderPackageRequest(ReadOnlySpan<byte> payload)
    {
        var reader = new PayloadReader(payload);
        var value = new UyaRenderPackageRequestPayload(
            reader.ReadString(), reader.ReadString(), reader.ReadString(), reader.ReadUInt32(),
            reader.ReadString(), reader.ReadString());
        reader.Complete();
        return value;
    }

    public static byte[] EncodeUyaRenderPackageResult(UyaRenderPackageResultPayload value)
    {
        var writer = new PayloadWriter();
        writer.WriteString(value.RootPath);
        writer.WriteString(value.CacheKey);
        writer.WriteStrings(value.TerrainPaths);
        writer.WriteBoolean(value.SkyPath is not null);
        if (value.SkyPath is not null) writer.WriteString(value.SkyPath);
        writer.WriteBoolean(value.Environment is not null);
        if (value.Environment is { } environment)
        {
            writer.WriteUInt32(environment.BackgroundRed);
            writer.WriteUInt32(environment.BackgroundGreen);
            writer.WriteUInt32(environment.BackgroundBlue);
            writer.WriteUInt32(environment.FogRed);
            writer.WriteUInt32(environment.FogGreen);
            writer.WriteUInt32(environment.FogBlue);
            writer.WriteSingle(environment.FogNearDistance);
            writer.WriteSingle(environment.FogFarDistance);
            writer.WriteSingle(environment.FogNearIntensity);
            writer.WriteSingle(environment.FogFarIntensity);
            writer.WriteSingle(environment.DeathHeight);
            writer.WriteBoolean(environment.IsSphericalWorld);
            writer.WriteSingle(environment.SphereCenterX);
            writer.WriteSingle(environment.SphereCenterY);
            writer.WriteSingle(environment.SphereCenterZ);
            writer.WriteSingle(environment.ShipPositionX);
            writer.WriteSingle(environment.ShipPositionY);
            writer.WriteSingle(environment.ShipPositionZ);
            writer.WriteSingle(environment.ShipRotationZ);
            writer.WriteInt32(environment.ShipPath);
            writer.WriteInt32(environment.ShipCameraCuboidStart);
            writer.WriteInt32(environment.ShipCameraCuboidEnd);
            writer.WriteInt32(environment.ChunkPlaneCount);
            writer.WriteInt32(environment.CoreSoundsCount);
        }
        writer.WriteUInt32(checked((uint)value.Assets.Count));
        foreach (var asset in value.Assets)
        {
            writer.WriteString(asset.AssetId);
            writer.WriteString(asset.Kind);
            writer.WriteBoolean(asset.Path is not null);
            if (asset.Path is not null) writer.WriteString(asset.Path);
            writer.WriteBoolean(asset.Error is not null);
            if (asset.Error is not null) writer.WriteString(asset.Error);
        }
        var octants = value.OcclusionOctants ?? [];
        writer.WriteUInt32(checked((uint)octants.Count));
        foreach (var octant in octants)
        {
            writer.WriteInt32(octant.X);
            writer.WriteInt32(octant.Y);
            writer.WriteInt32(octant.Z);
            writer.WriteInt32(octant.MaskIndex);
        }
        writer.WriteBoolean(value.CacheHit);
        return writer.ToArray();
    }

    public static UyaRenderPackageResultPayload DecodeUyaRenderPackageResult(ReadOnlySpan<byte> payload)
    {
        var reader = new PayloadReader(payload);
        var rootPath = reader.ReadString();
        var cacheKey = reader.ReadString();
        var terrainPaths = reader.ReadStrings();
        var skyPath = reader.ReadBoolean() ? reader.ReadString() : null;
        var environment = reader.ReadBoolean()
            ? new UyaRenderEnvironmentPayload(
                reader.ReadUInt32(), reader.ReadUInt32(), reader.ReadUInt32(),
                reader.ReadUInt32(), reader.ReadUInt32(), reader.ReadUInt32(),
                reader.ReadSingle(), reader.ReadSingle(), reader.ReadSingle(), reader.ReadSingle(),
                reader.ReadSingle(), reader.ReadBoolean(),
                reader.ReadSingle(), reader.ReadSingle(), reader.ReadSingle(),
                reader.ReadSingle(), reader.ReadSingle(), reader.ReadSingle(), reader.ReadSingle(),
                reader.ReadInt32(), reader.ReadInt32(), reader.ReadInt32(), reader.ReadInt32(), reader.ReadInt32())
            : null;
        var assetCount = reader.ReadUInt32();
        if (assetCount > 100_000) PayloadFormat.Malformed("Render asset list exceeds item limit");
        var assets = new UyaRenderAssetPayload[assetCount];
        for (var index = 0; index < assets.Length; index++)
            assets[index] = new(
                reader.ReadString(),
                reader.ReadString(),
                reader.ReadBoolean() ? reader.ReadString() : null,
                reader.ReadBoolean() ? reader.ReadString() : null);
        var octantCount = reader.ReadUInt32();
        if (octantCount > 1_000_000) PayloadFormat.Malformed("Occlusion octant list exceeds item limit");
        var octants = new UyaRenderOcclusionOctantPayload[octantCount];
        for (var index = 0; index < octants.Length; index++)
            octants[index] = new(reader.ReadInt32(), reader.ReadInt32(), reader.ReadInt32(), reader.ReadInt32());
        var value = new UyaRenderPackageResultPayload(
            rootPath, cacheKey, terrainPaths, skyPath, environment, assets, reader.ReadBoolean(), octants);
        reader.Complete();
        return value;
    }
}

File: src/main/bridge/RenderPayloadCodec.ts
import { malformed, PayloadReader, PayloadWriter } from './PayloadIO.ts';
import type { UyaRenderPackageRequest, UyaRenderPackageResult } from '../../types/BridgePayloads.js';

export function encodeUyaRenderPackageRequest(value: UyaRenderPackageRequest): Buffer {
  const writer = new PayloadWriter();
  writer.writeString(value.sourceIsoPath);
  writer.writeString(value.cacheRootPath);
  writer.writeString(value.fingerprint);
  writer.writeUInt32(value.level);
  writer.writeString(value.projectPath);
  writer.writeString(value.catalogRootPath);
  return writer.toBuffer();
}

export function decodeUyaRenderPackageRequest(payload: Uint8Array): UyaRenderPackageRequest {
  const reader = new PayloadReader(payload);
  const value = {
    sourceIsoPath: reader.readString(),
    cacheRootPath: reader.readString(),
    fingerprint: reader.readString(),
    level: reader.readUInt32(),
    projectPath: reader.readString(),
    catalogRootPath: reader.readString(),
  };
  reader.complete();
  return value;
}

export function encodeUyaRenderPackageResult(value: UyaRenderPackageResult): Buffer {
  const writer = new PayloadWriter();
  writer.writeString(value.rootPath);
  writer.writeString(value.cacheKey);
  writer.writeStrings(value.terrainPaths);
  writer.writeBoolean(value.skyPath !== undefined);
  if (value.skyPath !== undefined) writer.writeString(value.skyPath);
  writer.writeBoolean(value.environment !== undefined);
  if (value.environment) {
    value.environment.backgroundColor.forEach((channel) => writer.writeUInt32(channel));
    value.environment.fogColor.forEach((channel) => writer.writeUInt32(channel));
    writer.writeFloat32(value.environment.fogNearDistance);
    writer.writeFloat32(value.environment.fogFarDistance);
    writer.writeFloat32(value.environment.fogNearIntensity);
    writer.writeFloat32(value.environment.fogFarIntensity);
  }
  writer.writeUInt32(value.assets.length);
  value.assets.forEach((asset) => {
    writer.writeString(asset.assetId);
    writer.writeString(asset.kind);
    writer.writeBoolean(asset.path !== undefined);
    if (asset.path !== undefined) writer.writeString(asset.path);
    writer.writeBoolean(asset.error !== undefined);
    if (asset.error !== undefined) writer.writeString(asset.error);
  });
  writer.writeBoolean(value.cacheHit);
  return writer.toBuffer();
}

export function decodeUyaRenderPackageResult(payload: Uint8Array): UyaRenderPackageResult {
  const reader = new PayloadReader(payload);
  const rootPath = reader.readString();
  const cacheKey = reader.readString();
  const terrainPaths = reader.readStrings();
  const skyPath = reader.readBoolean() ? reader.readString() : undefined;
  const environment = reader.readBoolean() ? {
    backgroundColor: [reader.readUInt32(), reader.readUInt32(), reader.readUInt32()] as [number, number, number],
    fogColor: [reader.readUInt32(), reader.readUInt32(), reader.readUInt32()] as [number, number, number],
    fogNearDistance: reader.readFloat32(),
    fogFarDistance: reader.readFloat32(),
    fogNearIntensity: reader.readFloat32(),
    fogFarIntensity: reader.readFloat32(),
  } : undefined;
  const assetCount = reader.readUInt32();
  if (assetCount > 100_000) throw new Error('Render asset list exceeds item limit');
  const assets = Array.from({ length: assetCount }, () => {
    const assetId = reader.readString();
    const kind = reader.readString();
    if (kind !== 'moby' && kind !== 'tie' && kind !== 'shrub') malformed('Render asset kind is invalid');
    const asset: UyaRenderPackageResult['assets'][number] = { assetId, kind };
    if (reader.readBoolean()) asset.path = reader.readString();
    if (reader.readBoolean()) asset.error = reader.readString();
    return asset;
  });
  const value = { rootPath, cacheKey, terrainPaths, skyPath, environment, assets, cacheHit: reader.readBoolean() };
  reader.complete();
  return value;
}
